packages/fast-components-react-base/src/listbox/listbox.tsx

```diff
@@ -0,0 +1,314 @@
+                child.focus();
+
+                this.setState({
+                    focusIndex,
+                });
+
+                break;
+            }
+
+            focusIndex += adjustment;
+        }
+    }
+
+    /**
+     * Handle the keydown event of the root menu
+     */
+    private handleMenuKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
+        switch (e.keyCode) {
+            case KeyCodes.arrowDown:
+            case KeyCodes.arrowRight:
+                e.preventDefault();
+                this.setFocus(this.state.focusIndex + 1, 1);
+
+                break;
+
+            case KeyCodes.arrowUp:
+            case KeyCodes.arrowLeft:
+                e.preventDefault();
+                this.setFocus(this.state.focusIndex - 1, -1);
+
+                break;
+
+            case KeyCodes.end:
+                e.preventDefault();
+                this.setFocus(this.domChildren().length - 1, -1);
+
+                break;
+
+            case KeyCodes.home:
+                e.preventDefault();
+                this.setFocus(0, 1);
+
+                break;
+
+            default:
+                if (!e.ctrlKey) {
+                    this.processTypeAhead(e);
+                }
+        }
+    };
+
+    private processTypeAhead = (e: React.KeyboardEvent<HTMLDivElement>): void => {
+        const acceptedChars: string[] = [
+            "a",
+            "b",
+            "c",
+            "d",
+            "e",
+            "f",
+            "g",
+            "h",
+            "i",
+            "j",
+            "k",
+            "l",
+            "m",
+            "n",
+            "o",
+            "p",
+            "q",
+            "r",
+            "s",
+            "t",
+            "u",
+            "v",
+            "w",
+            "x",
+            "y",
+            "z",
+            "1",
+            "2",
+            "3",
+            "4",
+            "5",
+            "6",
+            "7",
+            "8",
+            "9",
+            "0",
+        ];
+
+        const newChar: string = e.key.toLowerCase();
+
+        if (acceptedChars.indexOf(newChar) === -1) {
+            return;
+        }
+
+        e.preventDefault();
+
+        clearTimeout(this.typeAheadTimer);
+
+        this.typeAheadString = this.typeAheadString + newChar;
+        const children: Element[] = this.domChildren();
+        let matchIndex: number = -1;
+        this.typeAheadData.some(
+            (typeAheadData: TypeAheadDataItem, index: number): boolean => {
+                if (typeAheadData.compareString.includes(this.typeAheadString)) {
+                    matchIndex = typeAheadData.index;
+                    return true;
+                }
+
+                return false;
+            }
+        );
+
+        if (matchIndex !== -1) {
+            this.typeAheadTimer = setTimeout((): void => {
+                this.typeAheadTimerExpired();
+            }, 1000);
+            this.setFocus(matchIndex, 1);
```

Comment:
can you explain why we need a timer? should we not just check to see if there is a new element to focus on every keystroke?

Reply:
That appears to be how native select works now.  If I have an option that gets focus when I rapidly type "a-b-c" it does not get focus if I type "a-b" pause for a few secs, then "c".  It looks to me like the keystroke buffer expires after time has passed.

Reply:
"Type-ahead is recommended for all listboxes, especially those with more than seven options:
Type a character: focus moves to the next item with a name that starts with the typed character.
Type multiple characters in **rapid succession**: focus moves to the next item with a name that starts with the string of characters typed."  
 https://www.w3.org/TR/wai-aria-practices-1.1/#Listbox

Reply:
Makes sense.